src/exceptions.h

```diff
@@ -19,108 +19,84 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	{
+		va_list ap;
+		va_start(ap, format);
+		m_s = VStringPrintf(format, ap);
+		va_end(ap);
+	}
 
-	virtual const char * what() const throw()
+	virtual const char * what() const noexcept
 	{
 		return m_s.c_str();
 	}
 protected:
 	std::string m_s;
 };
 
-class AlreadyExistsException : public BaseException {
-public:
-	AlreadyExistsException(const std::string &s): BaseException(s) {}
-};
-
-class VersionMismatchException : public BaseException {
-public:
-	VersionMismatchException(const std::string &s): BaseException(s) {}
-};
-
-class FileNotGoodException : public BaseException {
-public:
-	FileNotGoodException(const std::string &s): BaseException(s) {}
-};
-
-class DatabaseException : public BaseException {
-public:
-	DatabaseException(const std::string &s): BaseException(s) {}
-};
-
-class SerializationError : public BaseException {
-public:
-	SerializationError(const std::string &s): BaseException(s) {}
-};
-
-class PacketError : public BaseException {
-public:
-	PacketError(const std::string &s): BaseException(s) {}
-};
-
-class SettingNotFoundException : public BaseException {
-public:
-	SettingNotFoundException(const std::string &s): BaseException(s) {}
-};
-
-class ItemNotFoundException : public BaseException {
-public:
-	ItemNotFoundException(const std::string &s): BaseException(s) {}
-};
-
-class ServerError : public BaseException {
-public:
-	ServerError(const std::string &s): BaseException(s) {}
-};
-
-class ClientStateError : public BaseException {
-public:
-	ClientStateError(const std::string &s): BaseException(s) {}
-};
-
-class PrngException : public BaseException {
-public:
-	PrngException(const std::string &s): BaseException(s) {}
-};
-
-class ModError : public BaseException {
-public:
-	ModError(const std::string &s): BaseException(s) {}
-};
-
-
 /*
-	Some "old-style" interrupts:
-*/
-
-class InvalidNoiseParamsException : public BaseException {
-public:
-	InvalidNoiseParamsException():
-		BaseException("One or more noise parameters were invalid or require "
-			"too much memory")
-	{}
-
-	InvalidNoiseParamsException(const std::string &s):
-		BaseException(s)
-	{}
-};
-
-class InvalidPositionException : public BaseException
-{
-public:
-	InvalidPositionException():
-		BaseException("Somebody tried to get/set something in a nonexistent position.")
-	{}
-	InvalidPositionException(const std::string &s):
-		BaseException(s)
-	{}
-};
+ * Clang supports C++11 Constructor Delegation with printf-style constructors,
+ * but GCC doesn't. So this macro is required for now.
+ */
+#define DEFINE_EXCEPTION_BASE(ClassName, EXTRA_CONSTRUCTORS) \
+	class ClassName : public BaseException { \
+	public: \
+		ClassName(const std::string &s) noexcept : BaseException(s) { } \
+		ClassName(std::string &&s) noexcept : BaseException(std::move(s)) { } \
+		PRINTF_ATTR(2, 3) \
+		explicit ClassName(PRINTF_FORMAT_PREFIX const char *format, ...) noexcept : \
+			BaseException(std::string()) \
```

Comment:
can't this defer to BaseException with `std::forward`?

Reply:
Forwarding all arguments creates ambiguity with the other constructors.
Having the first arg be `const char* fmt` and forwarding the rest, breaks the printf formatting attribute.

Reply:
Even doing direct delegation:

`using BaseException::BaseException;`

Breaks the printf attributes (but only on GCC). I would definitely consider this a GCC bug. I'll see about filing an issue.

Reply:
Unfortunately it seems to be intentionally unsupported behavior:

https://github.com/gcc-mirror/gcc/blob/919822adc923b00e995e8a148bdb8115a794b47b/gcc/cp/call.cc#L9690